tachyon/math/matrix/matrix_utils.h

```diff
@@ -45,22 +45,22 @@ MakeCirculant(const Eigen::MatrixBase<ArgType>& arg) {
                                  CirculantFunctor<ArgType>(arg.derived()));
 }
 
-template <typename PackedPrimeField, typename Derived, typename PrimeField>
-std::vector<PackedPrimeField> PackRowHorizontally(
-    const Eigen::MatrixBase<Derived>& matrix, size_t row,
-    std::vector<PrimeField>& remaining_values) {
+template <typename PackedPrimeField, typename PrimeField>
+std::vector<PackedPrimeField*> PackRowHorizontally(
+    Eigen::Block<RowMajorMatrix<PrimeField>>& matrix, size_t row,
+    std::vector<PrimeField*>& remaining_values) {
   size_t num_packed = matrix.cols() / PackedPrimeField::N;
   size_t remaining_start_idx = num_packed * PackedPrimeField::N;
-  remaining_values =
-      base::CreateVector(matrix.cols() - remaining_start_idx,
-                         [row, remaining_start_idx, &matrix](size_t col) {
-                           return matrix(row, remaining_start_idx + col);
-                         });
+  remaining_values = base::CreateVector(
+      matrix.cols() - remaining_start_idx,
+      [row, remaining_start_idx, &matrix](size_t col) {
+        return reinterpret_cast<PrimeField*>(
+            matrix.data() + row * matrix.cols() + remaining_start_idx + col);
```

Comment:
What do you mean by "Also, add a comment that the starting column index of the Eigen::Block should be 0; otherwise, this won't work."? I was under the impression that Eigen::Blocks themselves act like matrices so their access always starts from (0,0)

Reply:
Form discussion: matrix.data() gives a pointer of the data at the beginning of the Eigen block, yet this pointer continues and spans across the entire parent matrix's set of values. Thus, this block should have the same columns as the parent it is deriving from (at least in RowMajor form) for the logic to work properly.